README.md

```diff
@@ -43,36 +43,38 @@ $ git clone https://github.com/iterative/example-gto.git
 $ cd example-gto
 ```
 
-### Versioning
+### Versioning an artifact
 
-To register a new artifact or a new version, use `gto register`. This is usually
-done to mark significant changes to the artifact (such as a release or a
-deprecation).
+Registering a version is usually done to mark significant changes to the
+artifact. To release a new version (including the very first one), use
+`gto tag`.
 
 ```console
-$ gto register awesome-model
-Created git tag 'awesome-model@v0.0.1' that registers a new version
+$ gto tag awesome-model --version v0.0.1
+Created git tag 'awesome-model@v0.0.1'
 ```
 
 <details summary="What happens under the hood?">
 
 GTO creates a special Git tag for the artifact version, in a standard format:
 `{artifact_name}@{version_number}`.
 
-The version is now associated to the current Git commit (`HEAD`). You can have
-several versions in a given commit, ordered by their automatic version numbers.
+The version is now associated to the current Git commit (`HEAD`). You can use
+another Git commit if you provide it's hexsha as an additional argument, like
+`$ gto tag awesome-model abc1234`.
 
 </details>
 
-### Assing a stage
+### Assigning a stage
 
-Assing an actionable stage for a specific artifact version with `gto assign`.
-Stages can mark it's readiness for a specific consumer. You can plug in a real
-downsteam system via CI/CD or web hooks. For example: redeploy an ML model.
+To assign an actionable stage for a specific artifact version use the same
+`gto tag` command. Stages can mark the artifact readiness for a specific
+consumer. You can plug in a real downsteam system via CI/CD or web hooks, e.g.
+to redeploy an ML model.
 
 ```console
-$ gto assign awesome-model prod
-Created git tag 'awesome-model#prod#1' that adds stage 'prod' to 'v0.0.1'
+$ gto tag awesome-model --version v0.0.1 --stage prod
+Created git tag 'awesome-model#prod#1'
```

Comment:
Could this error message specific the version as well?
E.g. `Created git tag 'awesome-model#prod#1' based on v0.0.1 of awesome-model`

Reply:
Also, since 1) we can't assign a stage without a version and 2) `gto tag awesome-model --version v0.0.1` is very different behavior than `gto tag awesome-model --version v0.0.1 --stage prod`

What do you think about a `--to-version` option when using `--stage`?

`gto tag awesome-model --stage prod --to-version v0.0.1`

Reply:
Looks better - easier to read the command. Do you think this should do both version registration and promotion then?
```
$ gto tag x3 --version v0.0.1 --stage prod
Created git tag 'x3@v0.0.1'
Created git tag 'x3#prod#1'
```

Reply:
I guess whether this makes sense or not depends on whether `gto` should allow promotion to a stage from a git ref OR if only a particular version should be allowed to be promoted to a stage.
I kind of assumed the latter.

Reply:
By default, if you try to assign a stage, but the version doesn't exist for the REF, GTO will register it for you. But if you supply `--skip-registration` git tag, then it won't. 

I.e. let's suppose there are no versions for `x4`:

```
$ gto tag x4 HEAD --stage prod
Created git tag 'x3@v0.0.1'
Created git tag 'x3#prod#1'

$ gto tag x4 HEAD^1 --stage prod --skip-registration
Created git tag 'x3#prod#1'
```

Looks like the difference between `--version` and `--to-version` is this: if you provide `REF` and `--to-version`, you will get an error, cause there should be only one way to specify the Git revision to reference with Git tag you'll create. While if you provide `REF` and `--version`, this will be OK: you register a `version` and reference `REF` with Git tag you create. That's why if you provide `REF`, `--version` and `--stage`, you can both register a version and make an assignment.

Reply:
Ok, I get the idea, but I am not sure if one command should be allowed to create/modify/delete more than one tag at a time.
I'll defer to others. cc @tapadipti

Reply:
- I think it's better to keep the commands simple and not allow doing 2 things (register version and assign stage) with one command.
- And having 2 similar flags (`--version` and `--to-version`) looks confusing to me.
- Also, why do we want to generalize this to `tag`/`untag`, instead of making the commands specifically about versions and stages? Do we see more kinds of tags being required in the future, and want to be able to cover those potential scenarios? Since versions and stages have special meanings for us and since we treat them separately (eg, we only display the latest stage per version but we display all versions per model), I think it's easier to understand and remember the commands if they are separate for the distinct functionalities. Simplifying the name/syntax is a good idea, but not sure if combining them into a single command is.

Reply:
> I think it's better to keep the commands simple and not allow doing 2 things (register version and assign stage) with one command.
```
# do you mean we shouldn't allow this:
$ gto tag x5 HEAD --stage prod
Created git tag 'x5@v0.0.1'
Created git tag 'x5#prod#1'

# or this?
$ gto tag x5 HEAD --stage prod --version 1.0.0
```

Reply:
> Also, why do we want to generalize this to tag/untag, instead of making the commands specifically about versions and stages?

From the above comments from @mnrozhkov and @alex000kim I assume it's too hard to memorize 5 different commands and syntaxes for a new user. I agree on this direction: GTO does very simple things and should be a lightweight tool, and having a big tough-to-read README with many commands is what I would like to avoid.

On the other hand, if unifying commands brings more confusion than value, let's not do those. Currently (without `--to-version` and `--from-version`) I think it's pretty straightforward though (maybe it's needed to be clarified in docs): if you supply `--stage`, you do an assignment, otherwise, you do a registration.

Reply:
Iterated on this with @mnrozhkov and eventually decided that
1. `tag` looks cool, but if you try to investigate how it works, it becomes confusing (different behavior with `--stage` and without `--stage`
2. `register` and `assign` are easier cause they separate use cases well
3. `deprecate` can unassign stages and deregister artifacts

I've updated the README and it looks good now: we have 3 commands only: `register`, `assign`, `deprecate`. Almost without breaking changes. Feel free to read and provide any more feedback re this.

Reply:
I've implemented this. So now:
- in CLI you have `register`, `assign`, `deprecate`.
- in API you have `register`, `unregister`, `assign`, `unassign`, `deprecate`

This is because Studio is the main user of API, and having separate commands for them is clearer.
For human in CLI we'll have `deprecate` unifying `deregister` and `unassign`, making things simpler, hopefully.

A side note: it looks like API and CLI in GTO diverge more and more over time.